geanylua/glspi_doc.c

```diff
@@ -92,34 +83,51 @@ static const gchar* doc_idx_to_filename(gint idx) {
 
 
 
-/* Actvate and focus the specified document */
+/* Switch to the specified document tab */
 static gint glspi_activate(lua_State* L)
 {
-	gint idx=-1;
-	if (lua_gettop(L)>0) {
-		if (lua_isnumber(L,1)) {
-			idx=(lua_tonumber(L,1));
-			if (idx<0) { /* Negative number refers to (absolute) GtkNotebook index */
-				idx=(0-idx)-1;
-				if (idx>=gtk_notebook_get_n_pages(NOTEBOOK)) { idx=-1;}
-			} else { /* A positive number refers to the geany->documents_array index */
-				idx=doc_idx_to_tab_idx(idx-1);
+	gint lua_idx = -1;
+	gint tab_idx = -1;
+
+	if (lua_gettop(L) > 0) {
+		if (lua_isnumber(L, 1)) {
+			lua_idx = (lua_tointeger(L, 1));
+
+			/* Note: lua_idx == 0 is undefined */
+			if (lua_idx < 0) {
+				/* Negative number refers to (absolute) GtkNotebook index */
+				tab_idx = (0 - lua_idx) - 1;
+
+				if (tab_idx >= gtk_notebook_get_n_pages(NOTEBOOK)) {
+					tab_idx = -1;
+				}
+			} else if (lua_idx > 0) {
+				/* Positive number refers to the geany->documents_array index */
+				gint doc_idx = lua_idx - 1;
+				if ((doc_idx < geany->documents_array->len) && documents[doc_idx]->is_valid) {
+					tab_idx = document_get_notebook_page(documents[doc_idx]);
+				}
+			}
+		} else if (lua_isstring(L, 1)) {
+			/* String is full document path */
+			gint doc_idx = filename_to_doc_idx(lua_tostring(L, 1));
+			if (doc_idx >= 0) {
+				tab_idx = document_get_notebook_page(documents[doc_idx]);
 			}
-
 		} else {
-			if (lua_isstring(L,1)) {
-				idx=doc_idx_to_tab_idx(filename_to_doc_idx(lua_tostring(L, 1)));
-			} else {
-				if (!lua_isnil(L,1)) { return FAIL_STR_OR_NUM_ARG(1); }
+			if (!lua_isnil(L, 1)) {
+				return FAIL_STR_OR_NUM_ARG(1);
```

Comment:
that could be the last `else if` instead of an `else { if }` while you're at it.  Just nitpicking here though

Reply:
I can change this pending an opinion about change of scope.